# Homework 5: SQL, FEC Data, and Small Donors

## Due: 11:59pm Tuesday, November 7


In this homework, we're going to explore the Federal Election
Commission's data on the money exchanged during the 2016 election.

This homework has two main parts:

1. Answering questions and computing descriptive statistics on the data
2. Conducting a hypothesis test

This is very similar to what you've done before in this class. However, in this
homework almost all of our computations will be done using SQL.

## Getting Started

For this assignment, you're going to use a popular cloud services provider: Heroku. This will give you some experience provisioning a database in the cloud and working on that database from your computer.

Since the free tier of Heroku's Postgres service limits users to 10,000 rows of data, we've provided a subset of the FEC dataset for you to work with.

If you're interested, you can download and load the entire dataset from
http://www.fec.gov/finance/disclosure/ftpdet.shtml.  It is about 4GB and contains around 24 million rows. (With Heroku and other cloud services, it is relatively straightforward to rent clusters of machines to work on much larger datasets.  In particular, it would be easy to rerun your analyses in this assignment on the full dataset.)

## If you've already done this for lab08, you do not have to follow these steps again. Otherwise you can get duplicate data in your database

### Provisioning the Postgres DB

1. Visit https://signup.heroku.com/postgres-home-button and sign up for an account
if you don't have one already.
2. Now, install the Heroku CLI: https://devcenter.heroku.com/articles/heroku-cli.
Then, run `heroku login` to log into Heroku from your CLI.
3. Now, visit https://dashboard.heroku.com/apps and click **New -> App**. Name the app
whatever you want.
4. You should be sent to the app details page. Click **Resources** in the navbar, then
in the **Add-on** search bar, type "Postgres". You should be able to select **Heroku
Postgres**. Make sure the free tier (**Hobby Dev**) is selected and click **Provision**. Now
you should see **Heroku Postgres :: Database** in your **Add-ons** list.

### Loading the data into the Heroku DB

(1) Run the lines below in your terminal to install necessary libraries. (If you're running the jupyter notebook in the ds100 environment, the following 3 commands need to be ran in ds100 environment as well)

    conda install -y psycopg2
    conda install -y postgresql
    pip install ipython-sql

(2) Click the **Heroku Postgres :: Database** link in your app's **Add-ons** list.

(3) In the **Heroku Data** page you got redirected to, you should see the name of your
database. You can click on the **Setting** panel, then scroll down to **Administration** and click **View Credentials**. These are the
credentials that allow you to connect to the database. The last entry of the list
contains a line that looks like:

        heroku pg:psql db_name --app app_name

 In your terminal, take that command and add "`< fec.sql`" to the end
 to get something like: (You need to run this command in the folder which contains the fec.sql file. The file comes with the zip file of this homework)  

### ** Make sure you only run this ONCE, if you did it for lab, don't do it for homework**

        heroku pg:psql db_name --app app_name < fec.sql
        
        
 Run that command. It will run the commands in `fec.sql`, which load the dataset into the database.
 Now you should be able to run the command without the "`< fec.sql`" to
 have a postgres prompt. Try typing "`\d+`" at the prompt. You should get
 something like:
     
        ds100-hw4-db::DATABASE=> \d+
                               List of relations
         Schema |           Name           |   Type   |     Owner
        --------+--------------------------+----------+----------------
         public | cand                     | table    | vibrgrsqevmzkj
         public | comm                     | table    | vibrgrsqevmzkj
         public | ds100grades              | table    | vibrgrsqevmzkj
         public | ds100grades_recordid_seq | sequence | vibrgrsqevmzkj
         public | ds100weights             | table    | vibrgrsqevmzkj
         public | indiv_sample             | table    | vibrgrsqevmzkj
         public | indiv_sample_top         | table    | vibrgrsqevmzkj
         public | students                 | table    | vibrgrsqevmzkj
        (8 rows)

Congrats! You now have a Postgres database running containing the data you need
for this project.

# Part 1: Descriptive Statistics

In [1]:
# Run this cell to set up your notebook

import numpy as np
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns
import sqlalchemy
sns.set()

!pip install -U okpy
from client.api.notebook import Notebook

Requirement already up-to-date: okpy in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages
Requirement already up-to-date: coverage==3.7.1 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from okpy)
Requirement already up-to-date: requests==2.12.4 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from okpy)


In [2]:
ok = Notebook('hw5.ok')
ok.auth(force=False) # Change False to True if you are getting errors authenticating

Assignment: hw5
OK, version v1.13.9

Successfully logged in as sungbin.andy.kang@berkeley.edu


Now, let's connect to your Postgres database. On your Heroku Postgres details,
look at the credentials for the database. Take the long URI in the credentials and
replace the portion of the code that reads `"replace_me"` with the URI.

It should start with `postgres://`.

In [3]:
my_URI = 'postgres://sjoetjhfsdfmwl:0ec54984a830b11a5504022032551ac8774212af39fb103873c6f69f8e4db7f2@ec2-107-22-162-82.compute-1.amazonaws.com:5432/db7134929j6re5'
%load_ext sql
%sql $my_URI
engine = sqlalchemy.create_engine(my_URI)
connection = engine.connect()

## Table Descriptions

Here is a list of the tables in the database.  Each table links to the documentation on the [FEC page](http://www.fec.gov/finance/disclosure/ftpdet.shtml) for the dataset.

Note that the table names here are slightly different from the ones in lecture. Consult the FEC page
for the descriptions of the tables to find out what the correspondence is.

- [`cand`](http://www.fec.gov/finance/disclosure/metadata/DataDictionaryCandidateMaster.shtml): Candidates table. Contains names and party affiliation.
- [`comm`](http://www.fec.gov/finance/disclosure/metadata/DataDictionaryCommitteeMaster.shtml): Committees table. Contains committee names and types.
- [`indiv_sample`](http://www.fec.gov/finance/disclosure/metadata/DataDictionaryContributionsbyIndividuals.shtml): Sample of individual contributions from Berkeley.
- [`indiv_sample_top`](http://www.fec.gov/finance/disclosure/metadata/DataDictionaryContributionsbyIndividuals.shtml): Sample of individual contributions from the top contributors at Berkeley.


## Writing SQL queries

You can write SQL directly in the notebook by using the `%sql` magic, as demonstrated in the next cell.

Be careful when doing this.
If you try to run a SQL query that returns a lot of rows (100k or more is a good rule of thumb)
your browser will probably crash.

This is why in this homework, we will strongly prefer using SQL as much as
possible, only materializing the SQL queries when they are small.

**Because of this, your queries should work even as the size of your
data goes into the terabyte range! This is the primary advantage of working
with SQL as opposed to only dataframes.**

In [4]:
# We use `LIMIT 5` to avoid displaying a huge table.
# Although our tables shouldn't get too large to display,
# this is generally good practice when working in the
# notebook environment.  Jupyter notebooks don't handle
# very large outputs well. 
%sql SELECT * from indiv_sample LIMIT 5

5 rows affected.


[('C00404525', 'N', 'YE', 'P    ', 201601050300036659, '15', 'IND', 'KURZWEIL, JACK', 'BERKELEY', 'CA', '94703', 'RETIRED', 'RETIRED', '10062015', 500, None, 'SA01271666597', 1043014, None, None, 2012820161261113255),
 ('C00577130', 'A', 'M6', 'P    ', 201608159022353532, '15E', 'IND', 'LENHOFF, BERNIE', 'BERKELEY', 'CA', '947031658', 'ARBORICULTURAL SPECIALTIES', 'MANAGER', '05042016', 15, 'C00401224', 'VPF7BPFVV37', 1094141, None, '* EARMARKED CONTRIBUTION: SEE BELOW', 4081720161312601930),
 ('C00193433', 'N', 'M8', None, 201508149000826883, '15', 'IND', 'MURRA, LAURA S MS.', 'BERKELEY', 'CA', '94708', 'NONE', 'NONE', '07012015', 60, None, '4356366', 1021594, None, None, 4081720151248947285),
 ('C00042366', 'A', '30G', 'P    ', 201612080200695092, '15E', 'IND', 'FISHER, BRIGITTE', 'BERKELEY', 'CA', '947052825', 'NONE', 'NOT EMPLOYED', '11042016', 50, 'C00401224', 'SA011717509239506', 1139429, None, '* EARMARKED CONTRIBUTION: SEE BELOW', 1060820170035181652),
 ('C00577130', 'A', 'M3', 'P    ', 201606169018323111, '15E', 'IND', 'RISING, TODD', 'BERKELEY', 'CA', '947021513', 'ADAM BEYDA, PSY.D', 'PSYCHOLOGICAL ASSISTANT', '02212016', 27, 'C00401224', 'VPF7BKE39K0', 1077916, None, '* EARMARKED CONTRIBUTION: SEE BELOW', 4063020161300916281)]

For longer queries, you can save your query into a string, then use it in the
`%sql` statement. The `$query` in the `%sql` statement pulls in the value in
the Python variable `query`.

In [5]:
query = '''
SELECT cand_id, cand_name
FROM cand
WHERE cand_pty_affiliation = 'DEM'
LIMIT 5
'''

%sql $query

5 rows affected.


[('H0CA03078', 'BERA, AMERISH'),
 ('H0CA15148', 'HONDA, MIKE'),
 ('H2AZ09019', 'SINEMA, KYRSTEN'),
 ('H2CA00120', 'BROWNLEY, JULIA'),
 ('H2CA06259', 'HUFFMAN, JARED')]

In addition, you can assign the SQL statement to a variable and then call `.DataFrame()` on it to get a Pandas DataFrame.  

However, it will often be more efficient to express your computation directly in SQL.  For this homework, we will be grading your SQL expressions so be sure to do all computation in SQL (unless otherwise requested).

In [6]:
res = %sql select * from cand limit 5
res_df = res.DataFrame()
res_df['cand_id']

5 rows affected.


0    H0CA03078
1    H0CA15148
2    H2AZ09019
3    H2CA00120
4    H2CA06259
Name: cand_id, dtype: object

### Question 1a

We are interested in finding the individuals from Berkeley that donated large sums. To begin to answer this question, we will look at the `indiv_sample` table. We'll find all the transactions that exceed \$3,000. However, if there are a lot of transactions like that, it might not be useful to list them all.  So before actually finding the transactions, find out how many such transactions there are. Use only SQL to compute the answer.

(It should be a table with a single column called **`count`** and a single entry, the number of transactions.)  

We will be grading the query string `query_q1a`.  You may modify our template but the result should contain the same information with the same names.

In [7]:
# complete the query string
query_q1a = """
SELECT COUNT(*)
FROM indiv_sample
WHERE transaction_amt > 3000
"""
q1a = %sql $query_q1a
q1a

1 rows affected.


[(4,)]

In [8]:
_ = ok.grade('q01a')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/6688xN
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 1b
Having seen that there aren't too many transactions that exceed \$3,000, let's find them all.  Using only SQL, construct a table containing the committee ID, contributor's name, and the transaction amount, for transactions that exceed $3,000 dollars.  Sort the transactions in decreasing order by amount. If two contributors contain the same transaction amount, sort by alphabetical order of their names.

We will be grading the query string `query_q1b`.  You may modify our template but the result should contain the same information with the same names.

In [9]:
%sql SELECT * from comm LIMIT 5

5 rows affected.


[('C00000729', 'AMERICAN DENTAL ASSOCIATION POLITICAL ACTION COMMITTEE', 'HARRISON, THOMAS C. DR', '1111 14TH STREET, NW', 'SUITE 1100', 'WASHINGTON', 'DC', '20005', 'B', 'Q', 'UNK', 'M', 'M', 'AMERICAN DENTAL ASSOCIATION', None),
 ('C00000935', 'DCCC', 'WARD, KELLY C.', '430 SOUTH CAPITOL STREET, SE', '2ND FLOOR', 'WASHINGTON', 'DC', '200034024', 'U', 'Y', 'DEM', 'M', None, None, None),
 ('C00001016', 'CARPENTERS LEGISLATIVE IMPROVEMENT COMMITTEE UNITED BROTHERHOOD OF CARPENTERS AND JOINERS', 'SILINS, ANDRIS', '101 CONSTIUTION AVENUE, NW', '10TH FLOOR WEST', 'WASHINGTON', 'DC', '20001', 'B', 'Q', 'NNE', 'M', 'L', 'UNITED BROTHERHOOD OF CARPENTERS AND JOINERS OF AMERICA', None),
 ('C00003418', 'REPUBLICAN NATIONAL COMMITTEE', 'PARKER, ANTHONY W MR', '310 FIRST STREET SE', None, 'WASHINGTON', 'DC', '20003', 'U', 'Y', 'REP', 'M', None, 'HOUSE MAJORITY TRUST', None),
 ('C00009936', 'AMERICAN FEDERATION OF GOVT. EMPL. POLITICAL ACTION COMMITTEE', 'GAGE, JOHN MR.', '80 F STREET, NW', None, 'WASHINGTON', 'DC', '20001', 'B', 'Q', 'UNK', 'M', 'L', "AMERICAN FEDERATION OF GOV'T EMPLOYEES", None)]

In [10]:
# complete the query string
query_q1b = """
SELECT 
    comm.cmte_id AS committee_id,
    indiv_sample.name AS name,
    indiv_sample.transaction_amt AS transaction_amt
FROM comm, indiv_sample
WHERE transaction_amt > 3000 AND
    comm.cmte_id = indiv_sample.cmte_id
ORDER BY transaction_amt DESC, name ASC
"""

q1b = %sql $query_q1b
q1b

4 rows affected.


[('C00586537', 'GUND, LOUISE LAIDLAW', 25000),
 ('C00000935', 'FIDDLER, JERRY', 10000),
 ('C00586537', 'LAPPEN, DAVID', 5000),
 ('C00401224', 'LUEVANO, ROSA', 5000)]

In [11]:
_ = ok.grade('q01b')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/7633yB
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 1c
Of course, only looking at individual transactions could be misleading, since each contributor can make several transactions. A more interesting question is: How much did each contributor give *in total*?  Find the total transaction amounts after grouping by the name.  This time, just use `LIMIT 20` to limit your results to the top 20 total donations. Sort the results in descending order by total contribution amount. Break ties by alphabetical order of name as in the previous question.

Do you see any names that are familiar?

We will be grading the query string `query_q1c`.  You may modify our template but the result should contain the same information with the same names.

In [12]:
# complete the query string
query_q1c = '''
SELECT 
    name AS name,
    SUM(transaction_amt) AS total_transaction_amt 
FROM indiv_sample 
GROUP BY name 
ORDER BY total_transaction_amt DESC, name ASC
LIMIT 20
'''
q1c = %sql $query_q1c
q1c

20 rows affected.


[('GUND, LOUISE LAIDLAW', 25000),
 ('FIDDLER, JERRY', 10000),
 ('SHENKER, SCOTT', 8000),
 ('BERLEKAMP, ELWYN', 5550),
 ('ABRAMS, DENISE', 5400),
 ('LAPPEN, DAVID', 5000),
 ('LUEVANO, ROSA', 5000),
 ('BERNHARDT, ANTHONY', 3200),
 ('HUFF, GERALD', 3200),
 ('BIRD, KAREN', 2700),
 ('EPSTEIN, BOB', 2700),
 ('HAHN, SOPHIE', 2700),
 ('JOSEPH, DAVID', 2700),
 ('LITTMANN, NICOLE', 2700),
 ('LOGAN, JONATHAN', 2700),
 ('REINIS, JONATHAN ROY', 2700),
 ('SEIBEL, PETER', 2700),
 ('SHAPIRO, CARL', 2700),
 ('TAYLOR, JEROME', 2700),
 ('WILNER, DAVID', 2700)]

In [13]:
_ = ok.grade('q01c')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/98rrAz
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 2a

Another interesting thing is to see what is the distribution of contributors by occupation? Let's first compute how many distinct occupations occur in the dataset. You may need to use the `DISTINCT` keyword to solve the problem.

In [14]:
query_q2a = '''
SELECT 
    COUNT(DISTINCT occupation)
FROM indiv_sample 
'''
q2a = %sql $query_q2a
q2a

1 rows affected.


[(407,)]

In [15]:
_ = ok.grade('q02a')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/0gRRoN
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 2b 
 Write a SQL query which computes for each occupation the number of transactions in the indiv_sample table. Display the top 20 results in descending order by count. 

We will be grading the query string `query_q2b`.  You may modify our template but the result should contain the same information with the same names.

In [16]:
# complete the query string
query_q2b = '''
SELECT 
    DISTINCT occupation AS occupation, 
    COUNT(*) AS count
FROM indiv_sample 
GROUP BY occupation
ORDER BY count DESC
LIMIT 20
'''
q2b = %sql $query_q2b
q2b

20 rows affected.


[('NOT EMPLOYED', 373),
 ('RETIRED', 306),
 (None, 70),
 ('PROFESSOR', 58),
 ('ATTORNEY', 56),
 ('NONE', 33),
 ('SCIENTIST', 30),
 ('ENGINEER', 20),
 ('PHONE CLERK', 20),
 ('RETIRED TEACHER', 20),
 ('ARTIST', 19),
 ('HOMEMAKER', 19),
 ('PSYCHOLOGIST', 19),
 ('PHYSICIAN', 18),
 ('UNEMPLOYED', 18),
 ('WRITER', 18),
 ('TEACHER', 17),
 ('LAWYER', 16),
 ('PSYCHOTHERAPIST', 16),
 ('PROFESSIONAL', 15)]

In [17]:
_ = ok.grade('q02b')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/g5JJDj
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Questions 2c
Compute the average transaction amount for each occupation, and sort them in descending order. Display the top 10 results.

In [18]:
# complete the query string
query_q2c = '''
SELECT 
    DISTINCT occupation AS occupation, 
    AVG(transaction_amt) AS avg_amount,
    COUNT(*) AS count
FROM indiv_sample 
GROUP BY occupation
ORDER BY avg_amount DESC
LIMIT 10
'''
q2c = %sql $query_q2c
q2c

10 rows affected.


[('LOUISE GUND FOUNDATION', Decimal('25000.000000000000'), 1),
 ('INVESTOR/BOARD MEMBER', Decimal('10000.0000000000000000'), 1),
 ('MARRIAGE & FAMILY THERAPIST', Decimal('5000.0000000000000000'), 1),
 ('CHAIRMAN', Decimal('2700.0000000000000000'), 1),
 ('INVESTMENT MANAGEMENT', Decimal('2700.0000000000000000'), 1),
 ('LIVE THEATRE PRODUCER', Decimal('2700.0000000000000000'), 1),
 ('AIDS CASE MANAGER', Decimal('2500.0000000000000000'), 1),
 ('WEALTH MANAGER', Decimal('1700.0000000000000000'), 1),
 ('INVESTOR', Decimal('1485.0000000000000000'), 2),
 ('ACCOUNTANT', Decimal('1475.0000000000000000'), 2)]

In [19]:
_ = ok.grade('q02c')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/j2RRg4
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Questions 2d
You may find that the top results only contains 1 or 2 transaction which are not useful. So we should only keep the reulsts that have more than 10 transcations. This time, you should use a nested `SELECT` to apply the filter on the results from q2c. After that, display the top 10 occupations that mave the largest average transaction amount.


In [20]:
# complete the query string
query_q2d = '''
SELECT 
    occupation AS occupation, 
    amount AS amount,
    count AS count
FROM
    (SELECT
         DISTINCT occupation AS occupation, 
         AVG(transaction_amt) AS amount,
         COUNT(*) AS count
     FROM indiv_sample 
     GROUP BY occupation) AS results
WHERE count > 10
ORDER BY amount DESC
LIMIT 10
'''
q2d = %sql $query_q2d
q2d

10 rows affected.


[('ATTORNEY', Decimal('423.9464285714285714'), 56),
 ('SOFTWARE ENGINEER', Decimal('371.0000000000000000'), 13),
 ('CONSULTANT', Decimal('343.0833333333333333'), 12),
 ('STUDENT', Decimal('317.5454545454545455'), 11),
 ('SCIENTIST', Decimal('298.3666666666666667'), 30),
 ('ENGINEER', Decimal('228.5000000000000000'), 20),
 ('ARTIST', Decimal('216.2631578947368421'), 19),
 ('PHYSICIAN', Decimal('193.8333333333333333'), 18),
 ('RETIRED', Decimal('143.7418300653594771'), 306),
 ('UNEMPLOYED', Decimal('140.8333333333333333'), 18)]

In [21]:
_ = ok.grade('q02d')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/lOYYk1
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Questions 2e
In the previous queries, we compute the average transaction amount. But since each contributor can make several transactions, this doesn't tell us much. In stead, we should compute the average amount that a contributor made for each type of occupation. To achieve this, let's first group the transactions by name to compute the total amount for each individual, which is similar to q1c. But this time, we should `SELECT` the occupation as well. Display the top 10 results order by total_amount.

In [22]:
# complete the query string
query_q2e = '''
SELECT 
    occupation AS occupation,
    name AS name,
    SUM(transaction_amt) AS total_amount
FROM indiv_sample
GROUP BY name, occupation
ORDER BY total_amount DESC
LIMIT 10
''' 
q2e = %sql $query_q2e
q2e

10 rows affected.


[('LOUISE GUND FOUNDATION', 'GUND, LOUISE LAIDLAW', 25000),
 ('INVESTOR/BOARD MEMBER', 'FIDDLER, JERRY', 10000),
 ('SCIENTIST', 'SHENKER, SCOTT', 8000),
 ('RETIRED', 'BERLEKAMP, ELWYN', 5550),
 ('ATTORNEY', 'ABRAMS, DENISE', 5400),
 ('NOT EMPLOYED', 'LUEVANO, ROSA', 5000),
 ('MARRIAGE & FAMILY THERAPIST', 'LAPPEN, DAVID', 5000),
 ('RETIRED', 'BERNHARDT, ANTHONY', 3200),
 ('SOFTWARE ENGINEER', 'HUFF, GERALD', 3200),
 ('STUDENT', 'JOSEPH, DAVID', 2700)]

In [23]:
_ = ok.grade('q02e')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/nZ55mp
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 2f
Let's continue with 2e, after computing the total amount for each individual. We can compute the average total_amount for each occupation type. Again, we should filter out the occupations that have small counts. Therefore, you should only select the top 10 occupations order by the average total_amount, with the constrains that the number of contributors in that occupation should be large than 5.


In [24]:
SELECT 
    DISTINCT occupation AS occupation,
    AVG(total_amount) AS avg_amount,
    count AS count
FROM
    (SELECT 
        occupation AS occupation,
        name AS name,
        SUM(transaction_amt) AS total_amount,
        COUNT(*) AS count
     FROM indiv_sample 
     GROUP BY name, occupation) AS result1
WHERE count > 5
GROUP BY occupation
ORDER BY avg_amount DESC
LIMIT 10

IndentationError: unexpected indent (<ipython-input-24-b51affcce072>, line 2)

In [25]:
# complete the query string
query_q2f = '''
SELECT 
    ... AS occupation,
    ... AS avg_amount,
    ... AS count
FROM
    (SELECT 
        occupation AS occupation
        
        ... 
     FROM
        (SELECT 
            occupation AS occupation
            SUM(transaction_amt) AS total_amount
         FROM indiv_sample 
         GROUP BY occupation 
        ) AS result1
        ...
        ...
    ) AS result2
...
...
'''
q2f = %sql $query_q2f
q2f

(psycopg2.ProgrammingError) syntax error at or near ".."
LINE 2:     ... AS occupation,
            ^
 [SQL: 'SELECT \n    ... AS occupation,\n    ... AS avg_amount,\n    ... AS count\nFROM\n    (SELECT \n        occupation AS occupation\n        \n        ... \n     FROM\n        (SELECT \n            occupation AS occupation\n            SUM(transaction_amt) AS total_amount\n         FROM indiv_sample \n         GROUP BY occupation \n        ) AS result1\n        ...\n        ...\n    ) AS result2\n...\n...']


In [26]:
_ = ok.grade('q02f')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
q02f > Suite 1 > Case 1

>>> func = lambda x:list(map(lambda y:(y[0],int(y[1]),y[2]),x))
>>> expected = [('SCIENTIST', 1118, 8), ('ATTORNEY', 593, 40), ('SOFTWARE ENGINEER', 535, 9), ('CONSULTANT', 374, 11), ('STUDENT', 349, 10), ('NOT-EMPLOYED', 331, 9), ('ARTIST', 273, 15), ('PHYSICIAN', 268, 13), ('ENGINEER', 253, 18), ('PROFESSOR', 240, 32)]
>>> func(connection.execute(query_q2f).fetchall()) == expected
Traceback (most recent call last):
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 939, in execute
    return self._execute_text(object, multiparams, params)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1097, in _execute_text
    statement, parameters
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/xGkkzP
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 3
Now let's examine the committees that appear in the `indiv_sample` table, with the information from the `comm` table. First let's take a look of the `comm` table.

In [27]:
%sql select * from comm limit 5

5 rows affected.


[('C00000729', 'AMERICAN DENTAL ASSOCIATION POLITICAL ACTION COMMITTEE', 'HARRISON, THOMAS C. DR', '1111 14TH STREET, NW', 'SUITE 1100', 'WASHINGTON', 'DC', '20005', 'B', 'Q', 'UNK', 'M', 'M', 'AMERICAN DENTAL ASSOCIATION', None),
 ('C00000935', 'DCCC', 'WARD, KELLY C.', '430 SOUTH CAPITOL STREET, SE', '2ND FLOOR', 'WASHINGTON', 'DC', '200034024', 'U', 'Y', 'DEM', 'M', None, None, None),
 ('C00001016', 'CARPENTERS LEGISLATIVE IMPROVEMENT COMMITTEE UNITED BROTHERHOOD OF CARPENTERS AND JOINERS', 'SILINS, ANDRIS', '101 CONSTIUTION AVENUE, NW', '10TH FLOOR WEST', 'WASHINGTON', 'DC', '20001', 'B', 'Q', 'NNE', 'M', 'L', 'UNITED BROTHERHOOD OF CARPENTERS AND JOINERS OF AMERICA', None),
 ('C00003418', 'REPUBLICAN NATIONAL COMMITTEE', 'PARKER, ANTHONY W MR', '310 FIRST STREET SE', None, 'WASHINGTON', 'DC', '20003', 'U', 'Y', 'REP', 'M', None, 'HOUSE MAJORITY TRUST', None),
 ('C00009936', 'AMERICAN FEDERATION OF GOVT. EMPL. POLITICAL ACTION COMMITTEE', 'GAGE, JOHN MR.', '80 F STREET, NW', None, 'WASHINGTON', 'DC', '20001', 'B', 'Q', 'UNK', 'M', 'L', "AMERICAN FEDERATION OF GOV'T EMPLOYEES", None)]

### Questions 3a
Let's first groupby the transactions in the `indiv_sample` table with `cmte_id` column (committee id), then count how many transactions and how much total amount of contribution there are for each `cmte_id`. Select the top 5 committee id order by the transactions count.

In [28]:
query_q3a='''
SELECT
    cmte_id AS committee_id,
    SUM(transaction_amt) AS total_amount,
    COUNT(*) AS count
FROM indiv_sample
GROUP BY committee_id
ORDER BY count DESC
LIMIT 5
'''
q3a = %sql $query_q3a
q3a


5 rows affected.


[('C00401224', 66228, 799),
 ('C00575795', 38002, 252),
 ('C00577130', 19920, 239),
 ('C00000935', 26040, 153),
 ('C00042366', 7271, 71)]

In [29]:
_ = ok.grade('q03a')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/y8PPAR
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Questions 3b
Only showing the committee id doesn't tell us much. Let's also select the committee information by joining with the `comm` table using the `cmite_id` field.

You will extend the query we have from q3a and then perform a `JOIN` with the `comm` table. Select extra committee information such as committee name, party_affiliation, city and state from `comm`. Select the top 10 committees order by transaction count.

In [30]:
query_q3b = '''
WITH cmte_count AS
(
  SELECT 
      cmte_id AS cmte_id,
      SUM(transaction_amt) AS total_amount,
      COUNT(*) AS count
  FROM indiv_sample 
  GROUP BY cmte_id
)
SELECT 
    cmte_nm AS cmte_name, 
    cmte_pty_affiliation AS party_affiliation,
    cmte_city AS city,
    cmte_st AS state,
    total_amount AS total_amount,
    count AS count
FROM comm c INNER JOIN cmte_count i ON c.cmte_id = i.cmte_id
ORDER BY count DESC
LIMIT 10
'''
q3b = %sql $query_q3b
q3b


10 rows affected.


[('ACTBLUE', None, 'SOMERVILLE', 'MA', 66228, 799),
 ('HILLARY FOR AMERICA', 'DEM', 'NEW YORK', 'NY', 38002, 252),
 ('BERNIE 2016', 'DEM', 'BURLINGTON', 'VT', 19920, 239),
 ('DCCC', 'DEM', 'WASHINGTON', 'DC', 26040, 153),
 ('DSCC', 'DEM', 'WASHINGTON', 'DC', 7271, 71),
 ('END CITIZENS UNITED', None, 'WASHINGTON', 'DC', 1129, 47),
 ('HILLARY VICTORY FUND', None, 'NEW YORK', 'NY', 44163, 45),
 ('MOVEON.ORG POLITICAL ACTION', None, 'WASHINGTON', 'DC', 2836, 37),
 ('CATHERINE CORTEZ MASTO FOR SENATE', 'DEM', 'LAS VEGAS', 'NV', 6878, 27),
 ("EMILY'S LIST", None, 'WASHINGTON', 'DC', 7988, 26)]

In [31]:
_ = ok.grade('q03b')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/AnDDXj
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 3c
From the results in q3b, it seems most of the party affiliation are DEM. So let's count the actual distribution of the party affiliation. Select all the different committee id from `indiv_sample`, joining with the `comm` table. Then group the results by party_affiliation and count how many committees there are for each party_affiliation. Order the results by committee count.

In [32]:
query_q3c = '''
WITH cmte AS
(
  SELECT 
      DISTINCT cmte_id AS cmte_id
  FROM indiv_sample 
)
SELECT 
    DISTINCT cmte_pty_affiliation AS party_affiliation,
    COUNT(*) AS count
FROM comm c JOIN cmte i ON c.cmte_id = i.cmte_id
GROUP BY party_affiliation
ORDER BY count DESC
'''
q3c = %sql $query_q3c
q3c


6 rows affected.


[('DEM', 58), (None, 44), ('REP', 8), ('UNK', 4), ('NNE', 2), ('GRE', 1)]

In [33]:
_ = ok.grade('q03c')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/BgBBvo
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



# Part 2: A Permutation Test in SQL

In this part, we're going to perform a hypothesis test using a permutation test. We'll perform nearly all our operations in SQL.

Here's our question: **Is there a difference in the proportion of donations made by small donors (under \$200) in Berkeley to Hillary Clinton's and Bernie Sanders' campaigns?**

**Probability Model:** To formulate a null hypothesis, we must consider the how the data are generated, i.e., we need to have a probability model for how the data arose.  We know that the sample was taken at random from the collection of all donations from Berkeley donors. We argue that Clinton attracts small donors at the same rate as Sanders. Then, given the proportion of donors in the combined sample of Clinton and Sanders donors, it's  just by chance that the proportion for Sanders and the proportion for Clinton don't match. 

With this model, we can figure out (approximate) the probability that the difference between these proportions would be as large or larger than the observed difference. To do this, we use a permuation test. 

**Permutation Test:**  In the permuation test, we simply mix up the small and large donors, and then tally the number of small donors to Clinton and the number to Sanders. We keep both the total number of donors to Clinton and the total number of donors to Sanders the same as in the sample, and we compare the two proportions by taking their difference. Essentially, we can think of this mixing as permuting the labels ("Clinton donor" and "Sanders donor") on all of the donors in our sample. For each random permutation, we tally the proportion of small donors to Clinton and Sanders and take their difference. After many permutations, we can examine the distribution of differences to see how likely it was to have observed our sample's difference. 

**Formal Hypothesis:** Now that we have a better understanding of the probability model, we formulate the null hypotheis as follows:

**$H_o$:** Clinton and Sanders attract the same proportion of small donors. Any observed difference in our sample is simply due to chance variation.

**$H_A$:**  The difference is real. 

Note: We might consider a *one-sided* alternative in this siutation. A one-sided alternative would be that Sanders' proportion is larger than Clinton's. Why would we do this?  Sanders made it part of his campaign that he was funded entirely by small donoations and that he wasn't obligated to `big money.' For this reason, we might be interested only in the side that shows Sanders having a greater proportion than Clinton, and not vice versa. 

For a review of permutation tests, we suggest looking over [this section on A/B testing](https://www.inferentialthinking.com/chapters/16/2/ab-testing.html) from the Data 8 textbook. Lecture #13 on Hypothesis testing includes an example of a permutation test. 


In this part, our skeletons will be minimal; it is your responsibility to use the right statements to answer the questions.

### Question 4

Before we begin, please think about and answer the following questions.

**For each question, state "Yes" or "No", followed by a one-sentence explanation.**

**(a)** If we were working with the entire FEC dataset instead of a sample,
would we still conduct a hypothesis test? Why or why not?

**(b)** Let's suppose we find that in our sample the difference in the proportion of money coming from
small donations by Berkeley donors to Hillary and Bernie's campaign is 0.3. Would
we still need to conduct a hypothesis test? Why or why not?

**(c)** Let's suppose we find that in our sample the difference in the proportion of money coming from
small donations by Berkeley donors to Hillary and Bernie's campaign is 0. Would
we still need to conduct a hypothesis test? Why or why not?


**(a)** No, we could look at the actual proportion for population if we had the entire FEC dataset.

**(b)** Yes, one sample might not reflect the entire population, so bootstrap sampling is required for certain confidence interval.

**(c)** Yes, for the same reason as (b), one sample showing the difference as 0 might just be an outlier.

### Question 5

As discussed, we've taken a simple random sample of 2000 rows of the original FEC data for individual contributions that only include contributions from Berkeley. This sample is stored in the table `indiv_sample`.

The individual contributions of donors are linked to committees, not candidates directly. Hillary's primary committee was called `HILLARY FOR AMERICA`, and Bernie's was `BERNIE 2016`.

Fill in the SQL query below to compute a SQL view called `contribs` of all the contributions for each
candidate's committee. Views are like tables. However, instead of storing the rows in the database, Postgres will recompute the values in the view each time you query it.

The resulting view should contain three columns:

1. `name`: The names of the each contributor
2. `transaction_amt`: The amount of each contribution
3. `cmte_nm`: The name of the committee that received the contribution (either `HILLARY FOR AMERICA` or `BERNIE 2016`).

Order your result by `transaction_amt` in descending order, breaking ties using `name` in alphabetical order.

We will be grading you on the query string `query_q5`.  You may modify our template but the result should contain the same information with the same names.

In [34]:
# Fill in the query
query_q5 = """
CREATE OR REPLACE TEMP VIEW contribs AS
SELECT name, transaction_amt, cmte_nm
FROM indiv_sample i JOIN comm c ON i.cmte_id = c.cmte_id
WHERE cmte_nm = 'HILLARY FOR AMERICA'
    OR cmte_nm = 'BERNIE 2016'
ORDER BY transaction_amt DESC, name
"""

%sql DROP VIEW IF EXISTS contribs CASCADE
%sql $query_q5
%sql SELECT * FROM contribs LIMIT 5

Done.
Done.
5 rows affected.


[('HAHN, SOPHIE', 2700, 'HILLARY FOR AMERICA'),
 ('JOSEPH, DAVID', 2700, 'HILLARY FOR AMERICA'),
 ('LITTMANN, NICOLE', 2700, 'BERNIE 2016'),
 ('LOGAN, JONATHAN', 2700, 'BERNIE 2016'),
 ('REINIS, JONATHAN ROY', 2700, 'HILLARY FOR AMERICA')]

In [35]:
_ = ok.grade('q05')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/Lg99Gv
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



You might notice that the maximum donation is \$2700. See if you can use Google to find out why.

### Question 6

Use the `sample` view to fill in the SQL query below to compute a table containing three columns:

- `cmte_nm`: The committee name
- `small`: The number of small donors to that committee
- `total`: The total number of donors to that committee

The resulting table should have this structure (your counts will not exactly match these):

<table>
    <tbody><tr>
        <th>cmte_nm</th>
        <th>small</th>
        <th>total</th>
    </tr>
    <tr>
        <td>BERNIE 2016</td>
        <td>1</td>
        <td>3</td>
    </tr>
    <tr>
        <td>HILLARY FOR AMERICA</td>
        <td>2</td>
        <td>4</td>
    </tr>
</tbody></table>

Then, uncomment the provided line of code to load the query into a DataFrame. Use the DataFrame to compute our observed statistic:

$$ \frac {\textrm{total donations under \$200 to Hillary}} {\textrm{total donations to Hillary}} - 
    \frac {\textrm{total donations under \$200 to Bernie}} {\textrm{total donations to Bernie}}. 
$$

Then, save the numerical value in the variable `observed_stat`.

Hint: We suggest computing the total amounts donated to each candidate using a WITH clause, then using the resulting table to compute the difference.

In [36]:
# Fill in the query
query_q6 = '''
WITH marked AS (
    SELECT *,
        CASE WHEN transaction_amt <= 200 THEN 1
            ELSE 0 END AS under
    FROM contribs
)
SELECT cmte_nm,
    SUM(under) AS small,
    COUNT(*) AS total
FROM marked
GROUP BY cmte_nm
ORDER BY cmte_nm
'''

# After writing query_q6, uncomment this line to create a dataframe
q6_df = pd.DataFrame(connection.execute(query_q6).fetchall(), columns=['cmte_nm', 'small', 'total'])


# Fill this in with the observed test statistic
observed_stat = (q6_df.loc[1][1] / q6_df.loc[1][2]) - (q6_df.loc[0][1] / q6_df.loc[0][2])

%sql $query_q6

2 rows affected.


[('BERNIE 2016', 212, 239), ('HILLARY FOR AMERICA', 225, 252)]

In [37]:
_ = ok.grade('q06')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/O7YYGE
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



Is this result surprising?  Even though the difference is quite small, this difference could still be considered large relative to the standard error of the statistic. We need to find the chance of a difference this large or larger under the null hypothesis.  That's where our permutation test comes into the picture.


### Conducting a Permutation Test

As mentioned above, in a permutation test, we generate new observations as follows: 

1. Shuffle the observed values between the two samples. In this case, shuffle the contributions between Hillary and Bernie. We do this because we assume that our null hypothesis is true and that it was just by chance that Clinton had a higher proportion of small donors than Sanders. 
2. Recompute the test statistic on the shuffled sample.
3. Repeat.

To do this in SQL, we need to number the rows we want to bootstrap.

The following cell creates a view called `sample`.

It adds a `row_id` column to each row in `contribs` corresponding to a contribution and removes the `name` column since we don't need it for the permutation test.

In [38]:
q_sample = '''
DROP VIEW IF EXISTS sample CASCADE;

CREATE OR REPLACE TEMP VIEW sample AS 
    SELECT row_number() over () AS row_id, transaction_amt, cmte_nm
    FROM contribs;

SELECT * FROM sample LIMIT 5
'''

connection.execute(q_sample)
%sql $q_sample

Done.
Done.
5 rows affected.


[(1, 2700, 'HILLARY FOR AMERICA'),
 (2, 2700, 'HILLARY FOR AMERICA'),
 (3, 2700, 'BERNIE 2016'),
 (4, 2700, 'BERNIE 2016'),
 (5, 2700, 'HILLARY FOR AMERICA')]

## Question 7

Let's contruct a view containing the shuffled `row_id`'s for each trial. For example, if we had 3 contributions total and wanted to create 100 shuffled samples, we want something that looks like:

```
trial_id | row_id
======== | ======
1        | 1
1        | 3
1        | 2
2        | 2
2        | 1
2        | 3
...      | ...
100      | 2
100      | 3
100      | 1
```

This will let us later construct a join on the `sample` view that computes the
shuffled sample for each trial.

Create a view called `design` that contains two columns: `trial_id`
and `row_id`. It should contain the IDs corresponding to
1000 reshuffles of the entire `sample` view. The `sample` view contains 491
rows, so the `design` view should have a total of
`1000 * 491 = 491000` rows.

Hint: See if you can generate the trial_id and row_id in sorted order first. Then, try to shuffle the rows within each `trial_id` using a clever ORDER BY clause. We've done something similar in class when we took a sample without replacement in SQL. Our solution uses the Postgres functions `generate_series` and `random()`.

In [39]:
# You may use these variables in your SQL query by prefixing the variable name with $
# eg. $n_trials
n_rows = 491
n_trials = 1000

query_q7 = """
CREATE VIEW design AS
SELECT trial_id.t AS trial_id, row_id.r AS row_id
    FROM generate_series(1, $n_trials) AS trial_id(t),
         generate_series(1, $n_rows) AS row_id(r)
ORDER BY trial_id, random()
"""

# Do not change anything below this line

# Fill in the $ variables set in the above string
import string
query_q7 = string.Template(query_q7).substitute(locals())

seed = 0.42
%sql drop view if exists design cascade
%sql SET SEED TO $seed
%sql $query_q7
%sql select * from design limit 5

Done.
Done.
Done.
5 rows affected.


[(1, 64), (1, 198), (1, 205), (1, 153), (1, 259)]

In [40]:
_ = ok.grade('q07')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/Y6EEZY
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 8

Construct a view called `joined` that uses the `sample` and `design` views to create a table containing the shuffled transaction amounts along with the **unshuffled** committee names — the committee names in the same order as they appeared in the original sample. Remember that in the permutation test we are shuffling the transaction amounts between the committees.

It should have three columns:

- `trial_id`: The number of the trial, from 1 to 10000
- `transaction_amt`: The shuffled transaction amounts
- `cmte_nm`: The **unshuffled** committee names.

It should look something like:

<table>
    <tbody><tr>
        <th>trial_id</th>
        <th>transaction_amt</th>
        <th>cmte_nm</th>
    </tr>
    <tr>
        <td>1</td>
        <td>10</td>
        <td>BERNIE 2016</td>
    </tr>
    <tr>
        <td>1</td>
        <td>25</td>
        <td>BERNIE 2016</td>
    </tr>
    <tr>
        <td>1</td>
        <td>25</td>
        <td>BERNIE 2016</td>
    </tr>
    <tr>
        <td>1</td>
        <td>25</td>
        <td>HILLARY FOR AMERICA</td>
    </tr>
    <tr>
        <td>1</td>
        <td>35</td>
        <td>HILLARY FOR AMERICA</td>
    </tr>
</tbody></table>

Hint: We've given you a WITH clause computing a row number for each trial group. Take a look at that table before trying out this problem. This problem requires you to join three tables together (one table gets joined twice).

In [41]:
query_q8 = '''
CREATE VIEW joined as
WITH num AS (
    SELECT MOD(row_number() over () - 1, $n_rows) + 1 AS row_in_group, *
    FROM design
),
     sample AS (
    SELECT row_number() over () AS row_id, transaction_amt, cmte_nm
    FROM contribs)

SELECT r.trial_id AS trial_id, a.transaction_amt AS transaction_amt, n.cmte_nm AS cmte_nm
FROM num r
    JOIN sample n
        ON n.row_id = r.row_in_group
    JOIN sample a
        ON a.row_id = n.row_id
ORDER BY trial_id, random()
'''

# Do not change anything below this line

query_q8 = string.Template(query_q8).substitute(locals())
%sql drop view if exists joined cascade
%sql SET SEED TO $seed
%sql $query_q8
%sql select * from joined limit 5

Done.
Done.
Done.
5 rows affected.


[(1, 25, 'HILLARY FOR AMERICA'),
 (1, 35, 'BERNIE 2016'),
 (1, 75, 'HILLARY FOR AMERICA'),
 (1, 7, 'HILLARY FOR AMERICA'),
 (1, 50, 'BERNIE 2016')]

In [42]:
_ = ok.grade('q08')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
q08 > Suite 1 > Case 1

>>> r=connection.execute(q_sample)
Traceback (most recent call last):
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 939, in execute
    return self._execute_text(object, multiparams, params)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1097, in _execute_text
    statement, parameters
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1121, in _execute_context
    None, None)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1402, in _handle_dbapi_exception
    exc_info
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/3l113p
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 9

Construct a query `query_q9` that uses the `joined` view to compute the number of small donors for each trial. Recall that a small donor is anyone who contributed \$200 or less.

It should have four columns:

- `trial_id`: The number of the trial, from 1 to 1000
- `cmte_nm`: The committee name
- `small`: The number of small donors for that committee, for that trial
- `total`: The number of total donors for that committee, for that trial

Order the result by increasing `trial_id` and then by `cmte_nm` within each trial.

Your result should look something like:

<table>
    <tbody><tr>
        <th>trial_id</th>
        <th>cmte_nm</th>
        <th>small</th>
        <th>total</th>
    </tr>
    <tr>
        <td>1</td>
        <td>BERNIE 2016</td>
        <td>207</td>
        <td>239</td>
    </tr>
    <tr>
        <td>1</td>
        <td>HILLARY FOR AMERICA</td>
        <td>230</td>
        <td>252</td>
    </tr>
    <tr>
        <td>2</td>
        <td>BERNIE 2016</td>
        <td>215</td>
        <td>239</td>
    </tr>
    <tr>
        <td>2</td>
        <td>HILLARY FOR AMERICA</td>
        <td>222</td>
        <td>252</td>
    </tr>
    <tr>
        <td>3</td>
        <td>BERNIE 2016</td>
        <td>216</td>
        <td>239</td>
    </tr>
</tbody></table>

Hint: This is similar to how we computed our observed test statistic above. We suggest first computing a column of 1s and 0s that indicate whether a contribution is small. Then, see if you can group that table properly.

Then, create a pandas DataFrame called `trials` that contains the `trial_id` for each trial as the index and the generated test statistics as its only column `stats`. It should look like:

<table class="dataframe" border="1">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>stats</th>
    </tr>
    <tr>
      <th>trial_id</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1</th>
      <td>-0.018629</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.030285</td>
    </tr>
    <tr>
      <th>3</th>
      <td>-0.002325</td>
    </tr>
    <tr>
      <th>4</th>
      <td>-0.010477</td>
    </tr>
    <tr>
      <th>5</th>
      <td>0.022133</td>
    </tr>
  </tbody>
</table>

In [43]:
# Fill in the query

query_q9 = '''
WITH marked AS (
    SELECT *,
        CASE WHEN transaction_amt <= 200 THEN 1
            ELSE 0 END AS under
    FROM joined
)
SELECT trial_id, cmte_nm, SUM(under), COUNT(*)
FROM marked
GROUP BY trial_id, cmte_nm
'''

# Do not change anything below this line

query_q9 = string.Template(query_q9).substitute(locals())
%sql SET SEED TO $seed
%sql $query_q9

# After writing query_q9, uncomment these line to create a dataframe
q9_df = pd.DataFrame(connection.execute(query_q9).fetchall(),
                      columns=['trial_id', 'cmte_nm', 'small', 'total'])

# Then, use q9_df to compute trials
q9_df["stats"] = q9_df["small"] / q9_df["total"]
trials = q9_df[q9_df["cmte_nm"] == "HILLARY FOR AMERICA"]['stats']
#- q9_df[q9_df["cmte_nm"] == "BERNIE 2016"]['stats']

trials
q9_df

Done.
2000 rows affected.


StatementError: (sqlalchemy.exc.ResourceClosedError) This Connection is closed [SQL: '\nWITH marked AS (\n    SELECT *,\n        CASE WHEN transaction_amt <= 200 THEN 1\n            ELSE 0 END AS under\n    FROM joined\n)\nSELECT trial_id, cmte_nm, SUM(under), COUNT(*)\nFROM marked\nGROUP BY trial_id, cmte_nm\n']

In [44]:
_ = ok.grade('q09')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
q09 > Suite 1 > Case 1

>>> r=connection.execute("SET SEED TO " + str(seed))
Traceback (most recent call last):
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 939, in execute
    return self._execute_text(object, multiparams, params)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1097, in _execute_text
    statement, parameters
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1121, in _execute_context
    None, None)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1402, in _handle_dbapi_exception
    exc_info
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203, i

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/7633MB
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Plotting the sample distribution

Run the following cell to make a plot of the distribution of the proportion differences between Hillary and Bernie. We draw the observed test statistic and the negative of the observed test statistic as red lines on the plot.

You should understand why we draw two red lines given our alternative hypothesis and our choice of test statistic.

In [45]:
sns.distplot(trials['stats'])
plt.axvline(x=observed_stat, color='r')
plt.axvline(x= -observed_stat, color='r')

NameError: name 'trials' is not defined

### Question 10

Write a query `query_q10` to compute the $p$-value based on the test statistics we generated through the shuffled samples.  We want to find the proportion of the 1000 test statistics that are at least as large in absolute value as the observed statistic.

Save your $p$-value in the variable `p_value`.

In [46]:
p_value = ...
p_value

Ellipsis

In [47]:
_ = ok.grade('q10')
_ = ok.backup()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
q10 > Suite 1 > Case 1

>>> 0 < p_value < 1
TypeError: '<' not supported between instances of 'int' and 'ellipsis'

# Error: expected
#     True
# but got
#     Traceback (most recent call last):
#       ...
#     TypeError: '<' not supported between instances of 'int' and 'ellipsis'

Run only this test case with "python3 ok -q q10 --suite 1 --case 1"
---------------------------------------------------------------------
Test summary
    Passed: 0
    Failed: 1
[k..........] 0.0% passed



<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Backup... 100% complete
Backup successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/backups/8q22gr
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



### Question 11

Based on your p-value, can you reject the null hypothesis using a cutoff of 5%? What do you conclude?

No, we do not reject the null hypothesis. The proportion of small donor is equivalent for both Hillary and Bernie.

How might the fact that the California primary is one of the last primaries (it's in June) impact the results?

It will not be affected in a significant way.

Congrats! You finished the homework.

# Submitting your assignment
First, run the next cell to run all the tests at once.

In [48]:
_ = ok.grade_all()

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
q01a > Suite 1 > Case 1

>>> connection.execute(query_q1a).fetchall() == [(4,)]
Traceback (most recent call last):
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 939, in execute
    return self._execute_text(object, multiparams, params)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1097, in _execute_text
    statement, parameters
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1121, in _execute_context
    None, None)
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1402, in _handle_dbapi_exception
    exc_info
  File "/home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 203

Now, run the cell below to submit your assignment to OkPy. The autograder should email you shortly with your autograded score. The autograder will only run once every 30 minutes.

**If you're failing tests on the autograder but pass them locally**, you should simulate the autograder by doing the following:

1. In the top menu, click Kernel -> Restart and Run all.
2. Run the cell above to run each OkPy test.

**You must make sure that you pass all the tests when running steps 1 and 2 in order.** If you are still failing autograder tests, you should double check your results.

In [49]:
# Now, we'll submit to okpy
_ = ok.submit()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Submit... 100% complete
Submission successful for user: sungbin.andy.kang@berkeley.edu
URL: https://okpy.org/cal/ds100/fa17/hw5/submissions/qjxxoy
NOTE: this is only a backup. To submit your assignment, use:
	python3 ok --submit



Now, run this cell to create a PDF to upload to Gradescope.

You may get an error about `wkhtmltopdf` the first time you run this cell. If so, install `wkhtmltopdf` and try again.

Do not modify the cell below.

In [51]:
!pip install -U gs100
from gs100 import convert
convert('hw5.ipynb', num_questions=3)

Requirement already up-to-date: gs100 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages
Requirement already up-to-date: pdfkit>=0.6 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from gs100)
Requirement already up-to-date: nbconvert>=5 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from gs100)
Requirement already up-to-date: nbformat>=4 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from gs100)
Requirement already up-to-date: beautifulsoup4>=4 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from gs100)
Requirement already up-to-date: toolz>=0.8 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from gs100)
Requirement already up-to-date: PyPDF2>=1.26 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from gs100)
Requirement already up-to-date: mistune>=0.7.4 in /home/ksab77/anaconda3/envs/ds100/lib/python3.6/site-packages (from nbconvert>=5->gs100)
Requirement already

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'hw5.ipynb'.
Created question_pdfs/q_email.pdf
Created question_pdfs/q04.pdf
Created question_pdfs/q11a.pdf
Created question_pdfs/q11b.pdf
Done! The resulting PDF is located in this directory and is called gradescope.pdf. Upload that PDF to Gradescope for grading.


**Make sure to upload your written answers to Gradescope now!**